### ISSUE TITLE
forecast of the battle results is not shown

### ISSUE THREAD

**Issue by MrZorG33** [We-the-People-civ4col-mod/Mod#557]
in the late game (**Plains** from 04/02/22) sometimes does not show the forecast of the results of the battle. so far I have only noticed this when attacking non-military ground units.
![image](https://user-images.githubusercontent.com/61798883/152968904-9304d008-b259-4b54-9172-b559cb532445.png)

however, the forecast of an attack on the Converted Natives is present.
![image](https://user-images.githubusercontent.com/61798883/152969651-1f88e75e-2052-4c07-8522-9ad12d32b545.png)

![image](https://user-images.githubusercontent.com/61798883/152969249-9b97525e-0358-4960-bb68-1f039c30fc7b.png)

and the lack of a forecast is not related to the number of units attacked.

**Comment by MrZorG33** [comment]
sometimes when loading a game, you may get this error:
![image](https://user-images.githubusercontent.com/61798883/152970819-bc11191f-5a34-4751-9124-1fc515609b7f.png)

Ray claims that this is possible due to the Russification of the game (but I did not translate this update).
and there is no folder on this path on my computer (game bought on Steam).
![image](https://user-images.githubusercontent.com/61798883/152972589-ef0607c1-14ed-4fd0-8a28-08e21cd8c671.png)

try loading this save on your computers (the error does not always appear).
[war.zip](https://github.com/We-the-People-civ4col-mod/Mod/files/8023020/war.zip)

this error appeared only when the battle forecast stopped showing.

**Comment by raystuttgart** [comment]
There is GameFonts for Russian Language / Chars.
The exception says it has a bug in "encoding a char".

It is a **"Unicode"** bug which has to be related to characters of a language.
So maybe it is not related to the Translation but it somehow related to a "non-english character".

See where it say:
**"UnicodeEncodeError"**

**Edit:**
It is **u0413 symbol** (a cyricllic symbol) --> see **message**
Just google it and it should tell the exact symbol it is.

**Edit 2:**
This is the symbol:
https://www.compart.com/de/unicode/U+0413

**Summary:**
This is definitely a bug that is specifically related to russian somehow.
(It might be triggered by a bug in the GameFonts. Currently do not know more.)

**Comment by raystuttgart** [comment]
**By the way:**
I really believe those are **2 separate bugs.**

**Bug 1:** Combat Preview not shown.
**Bug 2:** Python Exception related to encoding bug

----

But I will try to check it.

**Comment by raystuttgart** [comment]
@MrZorG33 
Please start a **fresh new game.**
Place those Units with Worldbuilder (just as I did) and check.

**Comment by MrZorG33** [comment]
@raystuttgart , once again: the path to the file, in which "supposedly" Cyrillic, does not exist. it is a virtual CD (I made a screenshot for this). the game is installed in a different way. why there is a reference to this symbol, I can not understand. but this has happened before.

also I believe that these two errors are related, since the second appeared immediately after the first. the previous save never caused Python errors.

Initially, this error did not exist and the forecast for the results of the battle was always there. so I think using the World Builder is not correct.

"_This is just about the "Combat Odds".
(I think this is already fixed in XML.)_"
I can try to inject the code into my current version to check. unless, of course, I have to change something else to do so.

**Comment by ErikSandbraaten** [comment]
In the third picture shown, can the attacker enter the plot where the stack of colonists are located ? If not, then the reason is that I added a check to the attacker so that the combat odds will only be shown if the attacker can enter the target tile.

**Comment by ErikSandbraaten** [comment]
I think that the issue could be resolved fairly quickly If you have a save where this can be reproduced

**Comment by MrZorG33** [comment]
> can the attacker enter the plot where the stack of colonists are located?

yes, he can enter this tile.
there is no save from that game, but such situations appear. I will try to attach the save as soon as I encounter this error again.

**Comment by MrZorG33** [comment]
I found this bug again (**develop2**). Screenshot and save attached.
[develop2_19.07.22.ColonizationSave.zip](https://github.com/We-the-People-civ4col-mod/Mod/files/9174445/develop2_19.07.22.ColonizationSave.zip)
![image](https://user-images.githubusercontent.com/61798883/180618863-0c8b6ff3-1fac-4a25-8986-c078e36eaa07.png)

**Comment by devolution79** [comment]
It's because of this check in CvUnit::canDefend:
'''
	if (getCapturingPlayer() != NO_PLAYER)
	{
		return false;
	}
'''

**Comment by devolution79** [comment]
Changing this is likely to have many side-effect so we may want to revert back to the commented out code in CvGameTextMgr::setCombatPlotHelp instead

```
bValid = gDLL->getInterfaceIFace()->getSelectionList()->canMoveInto(pPlot, true);
//bValid = !(pPlot->isWater());
```

But we'd have to ensure that it would work with Large Rivers (it won't unless we check for bridges etc.)

**Comment by raystuttgart** [comment]
By the way guys, I think this bug only existed in "develop" and does not exist in "New Hope" anymore.
(Since I remember to have had it fixed in XML of the Units.)

**Comment by raystuttgart** [comment]
**Edit:**
I just tested again in **"New Hope"** and I really think that this bug is fixed.
See my screenshot.

![Civ4ScreenShot0000](https://user-images.githubusercontent.com/59391125/206864725-33faf482-777c-4be8-a3ab-8b87f22fec38.JPG)

**Comment by raystuttgart** [comment]
Ok, I will close this unless it is reported again in which case we would reopen.

**Comment by MrZorG33** [comment]
Today, for the first time in a future release, New Hope encountered this bug.
unfortunately, on a laptop it is not possible to make a screen of the game - a screen of the desktop is taken. but you can try attacking a Free Colonist with a Dragoon.
[New Hope.zip](https://github.com/We-the-People-civ4col-mod/Mod/files/10268684/New.Hope.zip)

**Comment by raystuttgart** [comment]
A free Colonist with a Dragoon?
Ok I will try it. 👍

**Comment by raystuttgart** [comment]
No, it is working perfectly, see here:
![Civ4ScreenShot0002](https://user-images.githubusercontent.com/59391125/208679775-6ddce510-f3e3-4481-937e-6d7a534da258.JPG)

**Comment by raystuttgart** [comment]
Sorry, I just cannot reproduce this and so far you are the only user who is reporting this ...

**Comment by raystuttgart** [comment]
**Edit:**
I can reproduce it with your save now. 👍 
And I think In know the cause fo the issue.
I will try to fix it.

![Civ4ScreenShot0003](https://user-images.githubusercontent.com/59391125/208681694-088b798c-770d-4d86-8600-78de40b713f0.JPG)

**Comment by raystuttgart** [comment]
**Edit2:**
No, I could not find the issue in the code.
Maybe @Nightinggale could have a look as well.

**Comment by raystuttgart** [comment]
@MrZorG33 
Just to be 100% sure, did you do **any changes in XML** yourself?
While I am able to reproduce with your save, it is absolutely impossible for me to reproduce myself with WB.

**Comment by raystuttgart** [comment]
@devolution79 also had already analysed it but I do not get very smart of what should be done best.

It seems like it all starts with this here:

`CvGameTextMgr::setCombatPlotHelp`

```
bValid = gDLL->getInterfaceIFace()->getSelectionList()->canMoveInto(pPlot, true);
//bValid = !(pPlot->isWater());
```

**Comment by MrZorG33** [comment]
> Sorry, I just cannot reproduce this and so far you are the only user who is reporting this ... Did you modify anything in Civ4UnitInfos.xml?

No, I didn't make any changes. just updated the mod (most likely after your edits about the law and crime. but I think this should not affect the calculation of the result of the battle in any way).

**Comment by MrZorG33** [comment]
Ray, the problem is that many bugs can only be found during the game. and it is not necessary that they appear immediately. that's why I don't use WB in my tests, but start a new game after big updates.

**Comment by raystuttgart** [comment]
That is completely ok, but being able to reproduce it by WorldBuilder also means you understood the but systematically.
So meaning that I could not reproduce it means that I also have not understood the systematic cause yet.

**Comment by MrZorG33** [comment]
yes, like last time, this error is not permanent.

**Comment by devolution79** [comment]
I can look into this after I deal with the area stuff

**Comment by raystuttgart** [comment]
Awesome. 👍 
Thanks in advance.

**Comment by jo-oe** [comment]
There is one more interesting fact about the attack logic which is possibly linked with this bug:
See the setting in the save attached:
[testsetting.zip](https://github.com/We-the-People-civ4col-mod/Mod/files/10473648/testsetting.zip)
Player A has a stack of combat units A1, A2, Ax on plot P1
Player B has a stack of civilian units B1, B2, Bx on a neighbouring plot P2

option 1), correct behavior:
A1 attacks B1 and captures it, stays on P1
A2 attack B2 and captures it, stays on P1
Ax attacks Ax and captures it, moves to P2 (because no more enemy units on P2

option 2), bug:
stack A1-Ax attacks B1, captures B1 - ok
B2 to Bx are removed from the game - bug
A1 moves to P2 - ok
A2-Ax stay on P1, stack is split - also wrong, stack should imho move together

**Comment by raystuttgart** [comment]
@devolution79 
Would you have time to check this until end of February?
It would be great if we had one bug less on the list for the beta Release.

**Comment by MrZorG33** [comment]
_Today I encountered this error again. I am at war with the French and am trying to attack their Schooner with my Brig. as can be seen from the screenshot, an attack occurs (and not just a unit is moved), but the forecast of the results of the battle is not shown._

_upd._
_I first made this report and only then decided to continue the game. The French schooner is in Dutch territory (which I didn’t notice right away). could this affect the display of the forecast of the results of the battle?_

I did a little test and yes, this message can be considered erroneous, since the forecast was not shown due to the Schooner being in the territory of another European.
